xarray/core/dataset.py

```diff
@@ -3400,12 +3400,20 @@ def roll(self, roll_coords=None, **shifts):
             raise ValueError("dimensions %r do not exist" % invalid)
 
         variables = OrderedDict()
+        coord_names = set(k for k in self.coords if k in self.variables)
         for name, var in iteritems(self.variables):
-            if name in self.data_vars or roll_coords or roll_coords is None:
-                if roll_coords is None:
-                    warnings.warn("roll_coords will be set to False in the future."
-                                  " Explicitly set roll_coords to silence warning.",
-                                   DeprecationWarning, stacklevel=3)
+            if name in self.data_vars:
+                var_shifts = dict((k, v) for k, v in shifts.items()
+                                  if k in var.dims)
+                variables[name] = var.roll(**var_shifts)
+            elif name in coord_names and roll_coords:
+                var_shifts = dict((k, v) for k, v in shifts.items()
+                                  if k in var.dims)
+                variables[name] = var.roll(**var_shifts)
+            elif name in coord_names and roll_coords is None:
+                warnings.warn("roll_coords will be set to False in the future."
+                              " Explicitly set roll_coords to silence warning.",
+                               DeprecationWarning, stacklevel=3)
```

Comment:
My previous comment seemed confusing...
I think your previous commit is cleaner.

Or how about something like this?
```python
if roll_coords is None:
    warnings.warn("roll_coords will be set to False in the future."
                  " Explicitly set roll_coords to silence warning.",
                  DeprecationWarning, stacklevel=3)
    roll_coords = True

unrolled_vars = () if roll_coords else self.coord_names
variables = OrderedDict()
for k, v in self.variables:
    if k not in unrolled_vars:
        variables[k] = v.roll(**coords)
    else:
        variables[k] = v
```

Reply:
Yes, that definitely looks cleaner.